compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -100,13 +100,13 @@ object Implicits:
      */
     def companionRefs: TermRefSet = TermRefSet.empty
 
-    private var mySingletonClass: ClassSymbol = null
+    private var mySingletonClass: ClassSymbol | Null = null
 
     /** Widen type so that it is neither a singleton type nor a type that inherits from scala.Singleton. */
     private def widenSingleton(tp: Type)(using Context): Type = {
       if (mySingletonClass == null) mySingletonClass = defn.SingletonClass
       val wtp = tp.widenSingleton
-      if (wtp.derivesFrom(mySingletonClass)) defn.AnyType else wtp
+      if (wtp.derivesFrom(mySingletonClass.uncheckedNN)) defn.AnyType else wtp
```

Comment:
Is this safe in multi-thread?

Reply:
It looks safe to me. Once a variable is read as non-null in a thread, it will not revert back to null.

Reply:
If we never write null to a variable, could we make its type `ClassSymbol` instead of `ClassSymbol | Null` and initialize it to `uninitialized` instead of to `null`? A problem is that we would not be allowed to compare it to `null` (to test dynamically whether it is still uninitialized), but perhaps we can find or add a workaround. Maybe `if mySingletonClass == uninitialized` instead of `if mySingletonClass == null`, but that might be adding too much magic, proliferating `uninitialized` a bit too much.

Reply:
If there are replies to my comment above, please post them to https://github.com/lampepfl/dotty/issues/14622 instead of here.